service/node.go

```diff
@@ -1,25 +1,278 @@
 package service
 
 import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
 	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 
 	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/thecodeteam/csi-vsphere/pkg/vmware/fs"
+	"github.com/thecodeteam/gocsi"
+	"github.com/thecodeteam/gofsutil"
 )
 
 func (s *service) NodePublishVolume(
 	ctx context.Context,
 	req *csi.NodePublishVolumeRequest) (
 	*csi.NodePublishVolumeResponse, error) {
 
-	return nil, nil
+	var (
+		isBlock    bool
+		fsType     string
+		mntFlags   []string
+		readOnly   = req.Readonly
+		accessMode = req.VolumeCapability.AccessMode.Mode
+	)
+
+	lf := log.Fields{
+		"volumeID":   req.VolumeId,
+		"targetPath": req.TargetPath,
+		"readOnly":   req.Readonly,
+		"accessMode": accessMode,
+	}
+
+	switch accessType := req.VolumeCapability.AccessType.(type) {
+	case *csi.VolumeCapability_Block:
+		if accessType.Block == nil {
+			return nil, gocsi.ErrBlockTypeRequired
+		}
+		if readOnly {
+			return nil, status.Error(
+				codes.InvalidArgument,
+				"read only not supported by access type")
+		}
+		isBlock = true
+		lf["accessType"] = "block"
+	case *csi.VolumeCapability_Mount:
+		if accessType.Mount == nil {
+			return nil, gocsi.ErrMountTypeRequired
+		}
+		fsType = accessType.Mount.FsType
+		mntFlags = accessType.Mount.MountFlags
+		if readOnly {
+			hasROMntFlag := false
+			for _, o := range mntFlags {
+				if o == "ro" {
+					hasROMntFlag = true
+					break
+				}
+			}
+			if !hasROMntFlag {
+				mntFlags = append(mntFlags, "ro")
+			}
+		}
+		lf["accessType"] = "mount"
+		lf["fsType"] = fsType
+		lf["mntFlags"] = mntFlags
+	default:
+		return nil, gocsi.ErrAccessTypeRequired
+	}
+
+	// Attach the VMDK to this node.
+	dev, err := s.ops.Attach(req.VolumeId, nil)
```

Comment:
I'm not familiar with how `vmdkops` works here, but will it gracefully handle if the device is already attached? I'm wondering if there is any code that can be short-circuited if the device is already attached. Thought maybe that isn't necessary, as that is left to the idempotent provider.

Reply:
Hi @codenrhoden,

The backend ESXi component's `attach` call is idempotent.